<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Fiama Guevara">

<title>Metabarcoding Workshop - Core-Microbiome</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Metabarcoding Workshop</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text"><i class="fa-solid fa-home" aria-label="home"></i></span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-wed" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Wed</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-wed">    
        <li class="dropdown-header">Intro to metabarcoding (TBA)</li>
        <li>
    <a class="dropdown-item" href="./02_osc.html">
 <span class="dropdown-text">Intro to OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./03_qc-trim.html">
 <span class="dropdown-text">QC &amp; trimmming</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-thu" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Thu</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-thu">    
        <li class="dropdown-header">ASV/OTU calling and pipeline considerations (TBA)</li>
        <li>
    <a class="dropdown-item" href="./05_dada.html">
 <span class="dropdown-text">Calling ASVs with DADA2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./06_div.html">
 <span class="dropdown-text">Alpha &amp; beta diversity</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-fri" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Fri</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-fri">    
        <li class="dropdown-header">Network analysis (TBA)</li>
        <li class="dropdown-header">Core microbiome analysis (TBA)</li>
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-homework" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Homework</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-homework">    
        <li>
    <a class="dropdown-item" href="./homework/instructions.html">
 <span class="dropdown-text">Instructions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./homework/shell.html">
 <span class="dropdown-text">Unix Shell</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./homework/R.html">
 <span class="dropdown-text">R</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-reference-pages" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Reference pages</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-reference-pages">    
        <li class="dropdown-header">Literature (TBA)</li>
        <li class="dropdown-header">Metadata &amp; data submission (TBA)</li>
        <li>
    <a class="dropdown-item" href="./ref/ref_software.html">
 <span class="dropdown-text">Software management at OSC</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/metabar-ws24">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/metabar-ws24/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#defining-a-core-microbiome" id="toc-defining-a-core-microbiome" class="nav-link active" data-scroll-target="#defining-a-core-microbiome">Defining a Core Microbiome</a></li>
  <li><a href="#we-will-learn-two-approaches" id="toc-we-will-learn-two-approaches" class="nav-link" data-scroll-target="#we-will-learn-two-approaches"><span class="header-section-number">1</span> We will learn two approaches:</a>
  <ul class="collapse">
  <li><a href="#composition-relative-abundance" id="toc-composition-relative-abundance" class="nav-link" data-scroll-target="#composition-relative-abundance"><span class="header-section-number">1.1</span> Composition (relative abundance)</a></li>
  <li><a href="#membership-and-composition" id="toc-membership-and-composition" class="nav-link" data-scroll-target="#membership-and-composition"><span class="header-section-number">1.2</span> Membership and Composition</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Core-Microbiome</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Fiama Guevara </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
THIS PAGE IS STILL UNDER CONSTRUCTION
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<hr>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Library needed</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"phyloseq"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"microbiome"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
microbiome R package (microbiome.github.com)
    


 Copyright (C) 2011-2022 Leo Lahti, 
    Sudarshan Shetty et al. &lt;microbiome.github.io&gt;</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'microbiome'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:ggplot2':

    alpha</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:base':

    transform</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggvenn"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: dplyr</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: grid</code></pre>
</div>
</div>
<section id="defining-a-core-microbiome" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="defining-a-core-microbiome">Defining a Core Microbiome</h2>
<p>Let’s start <em>defining</em> what is considered a core microbiome. A <strong><em>core microbiome</em></strong> comprises the members shared among two or more microbial assemblages associated with a habitat.</p>
<p>There are several approaches to define a core microbiome<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>:</p>
<ol type="1">
<li>Membership (shared taxa)</li>
<li>Composition (relative abundance)</li>
<li>Phylogenetic and functional redundancy</li>
<li>Persistence (temporal and spatial)</li>
<li>Connectivity (co-variation within and shared communities)</li>
</ol>
</section>
<section id="we-will-learn-two-approaches" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="we-will-learn-two-approaches"><span class="header-section-number">1</span> We will learn two approaches:</h2>
<p>Start reading the <strong><code>phyloseq</code></strong> object containing your data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>ps.dataset <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"results/ps_fulldata/bac22rot_w_ASV.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Take a look of your metadata and the variables you want to analyse. We will focus on <strong>Rotation</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(ps.dataset)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           Year Location Rotation  Plot TRT Block Sampling.Date
T22NW102AB 2022    NWARS       CS 102AB   2   100    June, 2022
T22NW102C  2022    NWARS       CS  102C   7   100    June, 2022
T22NW103AB 2022    NWARS      CSW 103AB   3   100    June, 2022
T22NW103C  2022    NWARS      CSW  103C   8   100    June, 2022
T22NW201AB 2022    NWARS      CSW 201AB   3   200    June, 2022
T22NW201C  2022    NWARS      CSW  201C   8   200    June, 2022
T22NW203A  2022    NWARS       CS  203A   7   200    June, 2022
T22NW203BC 2022    NWARS       CS 203BC   2   200    June, 2022
T22NW305AB 2022    NWARS       CS 305AB   2   300    June, 2022
T22NW305C  2022    NWARS       CS  305C   7   300    June, 2022
T22NW403A  2022    NWARS       CS  403A   7   400    June, 2022
T22NW403BC 2022    NWARS       CS 403BC   2   400    June, 2022
T22NW404A  2022    NWARS      CSW  404A   8   400    June, 2022
T22NW404BC 2022    NWARS      CSW 404BC   3   400    June, 2022
T22W101AB  2022     WARS       CS 101AB   2   100    June, 2022
T22W101C   2022     WARS       CS  101C   7   100    June, 2022
T22W103AB  2022     WARS      CSW 103AB   3   100    June, 2022
T22W103C   2022     WARS      CSW  103C   8   100    June, 2022
T22W204A   2022     WARS      CSW  204A   8   200    June, 2022
T22W204BC  2022     WARS      CSW 204BC   3   200    June, 2022
T22W205A   2022     WARS       CS  205A   7   200    June, 2022
T22W205BC  2022     WARS       CS 205BC   2   200    June, 2022
T22W303AB  2022     WARS      CSW 303AB   3   300    June, 2022
T22W303C   2022     WARS      CSW  303C   8   300    June, 2022
T22W304AB  2022     WARS       CS 304AB   2   300    June, 2022
T22W304C   2022     WARS       CS  304C   7   300    June, 2022
T22W403AB  2022     WARS       CS 403AB   2   400    June, 2022
T22W403C   2022     WARS       CS  403C   7   400    June, 2022
T22W404A   2022     WARS      CSW  404A   8   400    June, 2022
T22W404BC  2022     WARS      CSW 404BC   3   400    June, 2022</code></pre>
</div>
</div>
<section id="composition-relative-abundance" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="composition-relative-abundance"><span class="header-section-number">1.1</span> Composition (relative abundance)</h3>
<ol type="1">
<li>Transform the ASV count table to compositional data (relative abundance). For that we will use the R package <strong><code>microbiome</code></strong> (previously installed) with the function <strong><code>transform</code></strong>.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>ps.dataset.cm <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(ps.dataset, <span class="st">"compositional"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Determine the core taxa by the variable of interest (<strong>Rotation</strong>).For this you will subset the data by <strong>Rotation</strong> variable.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>ps.CS <span class="ot">&lt;-</span> <span class="fu">subset_samples</span>(ps.dataset.cm, Rotation <span class="sc">==</span> <span class="st">"CS"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>ps.CSW <span class="ot">&lt;-</span> <span class="fu">subset_samples</span>(ps.dataset.cm, Rotation <span class="sc">==</span> <span class="st">"CSW"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, get the core taxa using the function <strong><code>core_members</code></strong> from the R package <strong><code>microbiome</code></strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>core.taxa.CS <span class="ot">&lt;-</span> <span class="fu">core_members</span>(ps.CS, <span class="at">detection =</span> <span class="fl">0.001</span>, <span class="at">prevalence =</span> <span class="dv">95</span><span class="sc">/</span><span class="dv">100</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>core.taxa.CSW <span class="ot">&lt;-</span> <span class="fu">core_members</span>(ps.CSW, <span class="at">detection =</span> <span class="fl">0.001</span>, <span class="at">prevalence =</span> <span class="dv">95</span><span class="sc">/</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
How to set detection and prevalence threasholds
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Detection</strong> (limit of detection) will be the minimum <em>relative abundance</em> value at which taxa will be selected. This can be set by the user or determined empirically based on sequencing depth.</p>
<p><strong>Prevalence</strong> is the proportion of samples that will be selected given a detection threshold.</p>
</div>
</div>
<details>
<summary>
<b>Example</b> (click here)
</summary>
<p>Consider this mock data for relative abundance of <em>Pseudomonas fluorescens</em> in 10 soil microbiome samples:</p>
<table class="table">
<thead>
<tr class="header">
<th>Sample</th>
<th style="text-align: left;">Relative abundance (%)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td style="text-align: left;">0.002</td>
</tr>
<tr class="even">
<td>3</td>
<td style="text-align: left;">0.49</td>
</tr>
<tr class="odd">
<td>4</td>
<td style="text-align: left;">7.8</td>
</tr>
<tr class="even">
<td>5</td>
<td style="text-align: left;">12.1</td>
</tr>
<tr class="odd">
<td>6</td>
<td style="text-align: left;">0.52</td>
</tr>
<tr class="even">
<td>7</td>
<td style="text-align: left;">0.2</td>
</tr>
<tr class="odd">
<td>8</td>
<td style="text-align: left;">0.06</td>
</tr>
<tr class="even">
<td>9</td>
<td style="text-align: left;">9.1</td>
</tr>
<tr class="odd">
<td>10</td>
<td style="text-align: left;">0.08</td>
</tr>
</tbody>
</table>
<p>Based on detection values, the prevalence will change. Lower detection threshold equals higher prevalence:</p>
<table class="table">
<thead>
<tr class="header">
<th>Detection</th>
<th style="text-align: left;">Prevalence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td style="text-align: left;">3/10</td>
</tr>
<tr class="even">
<td>0.1</td>
<td style="text-align: left;">6/10</td>
</tr>
<tr class="odd">
<td>0.01</td>
<td style="text-align: left;">9/10</td>
</tr>
</tbody>
</table>
</details>
<ol start="3" type="1">
<li>Create a list with the selected core taxa for both subsets</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>cores.list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">CS =</span> core.taxa.CS, <span class="at">CSW =</span> core.taxa.CSW)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>cores.list <span class="co">#To see the list of core taxa by Rotation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$CS
 [1] "ASV_1372" "ASV_1425" "ASV_1426" "ASV_1723" "ASV_1737" "ASV_2316"
 [7] "ASV_2440" "ASV_2450" "ASV_2483" "ASV_2545" "ASV_2605" "ASV_2635"
[13] "ASV_2825" "ASV_3076" "ASV_3576" "ASV_3587" "ASV_3592" "ASV_3986"
[19] "ASV_4073" "ASV_4082" "ASV_4242" "ASV_4348" "ASV_4391" "ASV_4478"
[25] "ASV_4504" "ASV_4555" "ASV_4565" "ASV_4642" "ASV_5779" "ASV_5791"
[31] "ASV_6609" "ASV_7160" "ASV_7731" "ASV_7736" "ASV_7822" "ASV_7847"
[37] "ASV_7848" "ASV_8079"

$CSW
 [1] "ASV_1367" "ASV_1372" "ASV_1388" "ASV_1425" "ASV_1426" "ASV_1723"
 [7] "ASV_1737" "ASV_1863" "ASV_2316" "ASV_2440" "ASV_2450" "ASV_2483"
[13] "ASV_2545" "ASV_2605" "ASV_2619" "ASV_2635" "ASV_2825" "ASV_2896"
[19] "ASV_3076" "ASV_3576" "ASV_3592" "ASV_3898" "ASV_3962" "ASV_3986"
[25] "ASV_4059" "ASV_4073" "ASV_4082" "ASV_4242" "ASV_4391" "ASV_4477"
[31] "ASV_4478" "ASV_4504" "ASV_4521" "ASV_4565" "ASV_4642" "ASV_4654"
[37] "ASV_4655" "ASV_5760" "ASV_5791" "ASV_6609" "ASV_6960" "ASV_7160"
[43] "ASV_7349" "ASV_7731" "ASV_7736" "ASV_7822" "ASV_7848" "ASV_7858"
[49] "ASV_8079"</code></pre>
</div>
</div>
<p>Get the list of shared taxa from the taxa table of your dataset</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>          Kingdom            Phylum               Class               Order
ASV_1372 Bacteria    Proteobacteria Alphaproteobacteria         Rhizobiales
ASV_1425 Bacteria    Proteobacteria Alphaproteobacteria         Rhizobiales
ASV_1426 Bacteria    Proteobacteria Alphaproteobacteria         Rhizobiales
ASV_1723 Bacteria    Proteobacteria Alphaproteobacteria    Sphingomonadales
ASV_1737 Bacteria    Proteobacteria Alphaproteobacteria    Sphingomonadales
ASV_2316 Bacteria    Proteobacteria Gammaproteobacteria     Xanthomonadales
ASV_2440 Bacteria    Proteobacteria Gammaproteobacteria     Burkholderiales
ASV_2450 Bacteria    Proteobacteria Gammaproteobacteria     Burkholderiales
ASV_2483 Bacteria    Proteobacteria Gammaproteobacteria     Burkholderiales
ASV_2545 Bacteria    Proteobacteria Gammaproteobacteria     Burkholderiales
ASV_2605 Bacteria    Proteobacteria Gammaproteobacteria     Burkholderiales
ASV_2635 Bacteria    Proteobacteria Gammaproteobacteria     Burkholderiales
ASV_2825 Bacteria       Myxococcota          Myxococcia        Myxococcales
ASV_3076 Bacteria       Myxococcota           Polyangia        Polyangiales
ASV_3576 Bacteria        Firmicutes             Bacilli          Bacillales
ASV_3592 Bacteria        Firmicutes             Bacilli          Bacillales
ASV_3986 Bacteria  Actinobacteriota     Thermoleophilia Solirubrobacterales
ASV_4073 Bacteria  Actinobacteriota     Thermoleophilia          Gaiellales
ASV_4082 Bacteria  Actinobacteriota     Thermoleophilia          Gaiellales
ASV_4242 Bacteria  Actinobacteriota      Actinobacteria          Frankiales
ASV_4391 Bacteria  Actinobacteriota      Actinobacteria   Micromonosporales
ASV_4478 Bacteria  Actinobacteriota      Actinobacteria       Micrococcales
ASV_4504 Bacteria  Actinobacteriota      Actinobacteria       Micrococcales
ASV_4565 Bacteria  Actinobacteriota      Actinobacteria   Pseudonocardiales
ASV_4642 Bacteria  Actinobacteriota      Actinobacteria Propionibacteriales
ASV_5791 Bacteria   Acidobacteriota      Acidobacteriae    Acidobacteriales
ASV_6609 Bacteria   Planctomycetota       Phycisphaerae    Tepidisphaerales
ASV_7160 Bacteria Verrucomicrobiota    Verrucomicrobiae  Chthoniobacterales
ASV_7731 Bacteria       Chloroflexi              KD4-96                &lt;NA&gt;
ASV_7736 Bacteria       Chloroflexi              KD4-96                &lt;NA&gt;
ASV_7822 Bacteria   Gemmatimonadota    Gemmatimonadetes    Gemmatimonadales
ASV_7848 Bacteria   Gemmatimonadota    Gemmatimonadetes    Gemmatimonadales
ASV_8079 Bacteria Methylomirabilota    Methylomirabilia     Rokubacteriales
                       Family                  Genus       Species
ASV_1372    Xanthobacteraceae           Pseudolabrys          &lt;NA&gt;
ASV_1425    Xanthobacteraceae         Bradyrhizobium       elkanii
ASV_1426    Xanthobacteraceae         Bradyrhizobium     japonicum
ASV_1723    Sphingomonadaceae           Sphingomonas          &lt;NA&gt;
ASV_1737    Sphingomonadaceae           Sphingomonas daechungensis
ASV_2316     Xanthomonadaceae             Arenimonas          &lt;NA&gt;
ASV_2440              SC-I-84                   &lt;NA&gt;          &lt;NA&gt;
ASV_2450              SC-I-84                   &lt;NA&gt;          &lt;NA&gt;
ASV_2483              TRA3-20                   &lt;NA&gt;          &lt;NA&gt;
ASV_2545    Nitrosomonadaceae                 mle1-7          &lt;NA&gt;
ASV_2605       Comamonadaceae          Piscinibacter          &lt;NA&gt;
ASV_2635       Comamonadaceae          Piscinibacter          &lt;NA&gt;
ASV_2825        Myxococcaceae             Archangium       gephyra
ASV_3076        Polyangiaceae          Aetherobacter          &lt;NA&gt;
ASV_3576          Bacillaceae               Bacillus     circulans
ASV_3592          Bacillaceae               Bacillus          &lt;NA&gt;
ASV_3986 Solirubrobacteraceae        Solirubrobacter          &lt;NA&gt;
ASV_4073                 &lt;NA&gt;                   &lt;NA&gt;          &lt;NA&gt;
ASV_4082          Gaiellaceae                Gaiella          &lt;NA&gt;
ASV_4242      Acidothermaceae           Acidothermus          &lt;NA&gt;
ASV_4391   Micromonosporaceae                   &lt;NA&gt;          &lt;NA&gt;
ASV_4478    Cellulomonadaceae           Cellulomonas          &lt;NA&gt;
ASV_4504   Intrasporangiaceae             Oryzihumus        terrae
ASV_4565   Pseudonocardiaceae         Pseudonocardia          &lt;NA&gt;
ASV_4642      Nocardioidaceae           Nocardioides          &lt;NA&gt;
ASV_5791                 &lt;NA&gt;                   &lt;NA&gt;          &lt;NA&gt;
ASV_6609    WD2101 soil group                   &lt;NA&gt;          &lt;NA&gt;
ASV_7160  Chthoniobacteraceae Candidatus Udaeobacter          &lt;NA&gt;
ASV_7731                 &lt;NA&gt;                   &lt;NA&gt;          &lt;NA&gt;
ASV_7736                 &lt;NA&gt;                   &lt;NA&gt;          &lt;NA&gt;
ASV_7822    Gemmatimonadaceae           Gemmatimonas          &lt;NA&gt;
ASV_7848    Gemmatimonadaceae           Gemmatimonas          &lt;NA&gt;
ASV_8079                 &lt;NA&gt;                   &lt;NA&gt;          &lt;NA&gt;
         Species_exact confidence
ASV_1372          &lt;NA&gt;       0.80
ASV_1425          &lt;NA&gt;       0.95
ASV_1426          &lt;NA&gt;       0.57
ASV_1723          &lt;NA&gt;       0.99
ASV_1737 daechungensis       0.81
ASV_2316          &lt;NA&gt;       1.00
ASV_2440          &lt;NA&gt;       1.00
ASV_2450          &lt;NA&gt;       1.00
ASV_2483          &lt;NA&gt;       0.99
ASV_2545          &lt;NA&gt;       0.99
ASV_2605          &lt;NA&gt;       0.90
ASV_2635          &lt;NA&gt;       0.82
ASV_2825          &lt;NA&gt;       0.65
ASV_3076         rufus       0.94
ASV_3576          &lt;NA&gt;       0.63
ASV_3592          &lt;NA&gt;       0.73
ASV_3986          &lt;NA&gt;       1.00
ASV_4073          &lt;NA&gt;       1.00
ASV_4082          &lt;NA&gt;       1.00
ASV_4242          &lt;NA&gt;       1.00
ASV_4391          &lt;NA&gt;       1.00
ASV_4478          &lt;NA&gt;       0.97
ASV_4504          &lt;NA&gt;       0.57
ASV_4565          &lt;NA&gt;       0.95
ASV_4642          &lt;NA&gt;       1.00
ASV_5791          &lt;NA&gt;       1.00
ASV_6609          &lt;NA&gt;       1.00
ASV_7160          &lt;NA&gt;       1.00
ASV_7731          &lt;NA&gt;       1.00
ASV_7736          &lt;NA&gt;       1.00
ASV_7822          &lt;NA&gt;       0.99
ASV_7848          &lt;NA&gt;       1.00
ASV_8079          &lt;NA&gt;       1.00</code></pre>
</div>
</div>
<p>Get the list of unique taxa in Rotation CS</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>          Kingdom           Phylum            Class             Order
ASV_3587 Bacteria       Firmicutes          Bacilli        Bacillales
ASV_4348 Bacteria Actinobacteriota   Actinobacteria Micromonosporales
ASV_4555 Bacteria Actinobacteriota   Actinobacteria Pseudonocardiales
ASV_5779 Bacteria  Acidobacteriota   Acidobacteriae  Acidobacteriales
ASV_7847 Bacteria  Gemmatimonadota Gemmatimonadetes  Gemmatimonadales
                     Family         Genus      Species Species_exact confidence
ASV_3587        Bacillaceae      Bacillus  aryabhattai          &lt;NA&gt;       0.96
ASV_4348 Micromonosporaceae          &lt;NA&gt;         &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_4555 Pseudonocardiaceae Amycolatopsis mediterranei          &lt;NA&gt;       1.00
ASV_5779               &lt;NA&gt;          &lt;NA&gt;         &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_7847  Gemmatimonadaceae  Gemmatimonas         &lt;NA&gt;          &lt;NA&gt;       1.00</code></pre>
</div>
</div>
<p>Get the list of unique taxa in Rotation CSW</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>          Kingdom            Phylum               Class               Order
ASV_1367 Bacteria    Proteobacteria Alphaproteobacteria         Rhizobiales
ASV_1388 Bacteria    Proteobacteria Alphaproteobacteria         Rhizobiales
ASV_1863 Bacteria    Proteobacteria Alphaproteobacteria     Acetobacterales
ASV_2619 Bacteria    Proteobacteria Gammaproteobacteria     Burkholderiales
ASV_2896 Bacteria       Myxococcota           Polyangia        Polyangiales
ASV_3898 Bacteria  Actinobacteriota     Thermoleophilia Solirubrobacterales
ASV_3962 Bacteria  Actinobacteriota     Thermoleophilia Solirubrobacterales
ASV_4059 Bacteria  Actinobacteriota     Thermoleophilia          Gaiellales
ASV_4477 Bacteria  Actinobacteriota      Actinobacteria       Micrococcales
ASV_4521 Bacteria  Actinobacteriota      Actinobacteria       Micrococcales
ASV_4654 Bacteria  Actinobacteriota      Actinobacteria Propionibacteriales
ASV_4655 Bacteria  Actinobacteriota      Actinobacteria Propionibacteriales
ASV_5760 Bacteria   Acidobacteriota      Acidobacteriae    Acidobacteriales
ASV_6960 Bacteria Verrucomicrobiota    Verrucomicrobiae      Pedosphaerales
ASV_7349 Bacteria   Acidobacteriota    Vicinamibacteria  Vicinamibacterales
ASV_7858 Bacteria   Gemmatimonadota    Gemmatimonadetes    Gemmatimonadales
                    Family          Genus Species Species_exact confidence
ASV_1367 Xanthobacteraceae           &lt;NA&gt;    &lt;NA&gt;          &lt;NA&gt;       0.97
ASV_1388 Xanthobacteraceae           &lt;NA&gt;    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_1863  Acetobacteraceae           &lt;NA&gt;    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_2619    Comamonadaceae     Acidovorax  avenae          &lt;NA&gt;       1.00
ASV_2896           BIrii41           &lt;NA&gt;    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_3898             67-14           &lt;NA&gt;    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_3962             67-14           &lt;NA&gt;    &lt;NA&gt;          &lt;NA&gt;       0.96
ASV_4059              &lt;NA&gt;           &lt;NA&gt;    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_4477 Cellulomonadaceae   Cellulomonas    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_4521    Micrococcaceae   Arthrobacter    &lt;NA&gt;          &lt;NA&gt;       0.82
ASV_4654   Nocardioidaceae      Kribbella    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_4655   Nocardioidaceae      Kribbella    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_5760              &lt;NA&gt;           &lt;NA&gt;    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_6960   Pedosphaeraceae ADurb.Bin063-1    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_7349              &lt;NA&gt;           &lt;NA&gt;    &lt;NA&gt;          &lt;NA&gt;       1.00
ASV_7858 Gemmatimonadaceae   Gemmatimonas    &lt;NA&gt;          &lt;NA&gt;       1.00</code></pre>
</div>
</div>
<ol start="4" type="1">
<li>Plot a Venn Diagram</li>
</ol>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="09_core_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="membership-and-composition" class="level3" data-number="1.2">
<h3 data-number="1.2" class="anchored" data-anchor-id="membership-and-composition"><span class="header-section-number">1.2</span> Membership and Composition</h3>
<p>This approach takes into account presence/absence and relative abundance to select shared taxa. The threshold could arbitrary be set but here we will follow the 95th percentile approach[^2]. That means that we will assign core membership to taxa that fell within the 95th percentile of the ranked frequency and ranked relative abundance within the whole data.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Side note: 95th Percentile approach. <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Diagram of how the 95th percentile frequency and abundance core taxa were identified. The distribution of total relative abundance and frequency of taxa is assessed and the those ranked within the 95th percentile (top 5%) are selected as core.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/Percentile.95.approach.Cantoran2023.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></p>
</figure>
</div>
</div>
</div>
</div>
<ol type="1">
<li>Create a data frame from the <strong><code>phyloseq</code></strong> object</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">psmelt</span>(ps.dataset.cm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Take the total Relative Abundance sum of each genus</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>df.RA <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(df<span class="sc">$</span>Abundance, <span class="at">by =</span> <span class="fu">list</span>(<span class="at">Genus =</span> df<span class="sc">$</span>Genus), <span class="at">FUN=</span>sum) </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>df.RA <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">rename</span>(df.RA, <span class="at">TotalRA =</span> <span class="st">`</span><span class="at">x</span><span class="st">`</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="3" type="1">
<li>Determining 95th quantile of the RA data</li>
</ol>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>      95% 
0.1663236 </code></pre>
</div>
</div>
<ol start="4" type="1">
<li>Selecting for genera that fit the 95th cutoff</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>taxa<span class="fl">.95</span> <span class="ot">&lt;-</span> <span class="fu">subset</span>(df.RA, TotalRA <span class="sc">&gt;=</span> quantile.<span class="fl">95.</span>abund) </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>taxa.<span class="fl">95.</span>genus <span class="ot">&lt;-</span> taxa<span class="fl">.95</span><span class="sc">$</span>Genus</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="5" type="1">
<li>Determine the frequency 95th cutoff. We go back and use the original phyloseq object and agglomerate the data by Genus tax rank (it could be any tax rank you want e.g., Phylum, Family)</li>
</ol>
<ol start="6" type="1">
<li>Create a table of presence/abundance of ASVs. We will use 1 if present, 0 if not.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>df.PA <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">*</span>((df.ASVs.genus<span class="sc">&gt;</span><span class="dv">0</span>)<span class="sc">==</span><span class="dv">1</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>Occupancy <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(df.PA)<span class="sc">/</span><span class="fu">ncol</span>(df.PA) <span class="co">#Calculate occupancy</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>df.Freq <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(df.PA, Occupancy))</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>df.Freq <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(df.Freq, <span class="st">'Genus'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="7" type="1">
<li>Determining 95th quantile of the presence/abundance table</li>
</ol>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>95% 
  1 </code></pre>
</div>
</div>
<ol start="8" type="1">
<li>Select the genera that fit the 95th cutoff</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>Freq<span class="fl">.95</span> <span class="ot">&lt;-</span> <span class="fu">subset</span>(df.Freq, Occupancy <span class="sc">&gt;=</span> quantile.<span class="fl">95.</span>freq) <span class="co">#select cutoff </span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>Freq.<span class="fl">95.</span>genus <span class="ot">&lt;-</span> Freq<span class="fl">.95</span><span class="sc">$</span>Genus</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>Quantile<span class="fl">.95</span> <span class="ot">&lt;-</span> <span class="fu">merge.data.frame</span>(Freq<span class="fl">.95</span>, taxa<span class="fl">.95</span>, <span class="at">by =</span> <span class="st">"Genus"</span>)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Side note: You can plot a Venn Diagram to see how many taxa made the cutoffs. <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Use the <strong>Core.list</strong> data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>Core.<span class="fl">95.</span>list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Frequency =</span> Freq.<span class="fl">95.</span>genus, <span class="at">Relative_abundance =</span> taxa.<span class="fl">95.</span>genus)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>plot.VennD <span class="ot">&lt;-</span> <span class="fu">ggvenn</span>(Core.<span class="fl">95.</span>list,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">stroke_size =</span> <span class="fl">0.5</span>, <span class="at">set_name_size =</span> <span class="fl">3.2</span>, <span class="at">text_size =</span> <span class="dv">3</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(plot.VennD)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="09_core_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<ol start="9" type="1">
<li>Get the list of your core taxa based on the 95th percentile</li>
</ol>
<ol start="10" type="1">
<li>Subset your original phyloseq object for core taxa</li>
</ol>
<ol start="11" type="1">
<li>Fix the relative abundance of the core taxa to continue with the plot</li>
</ol>
<ol start="12" type="1">
<li>Create a bubble plot You can customize your plot using <em><code>ggplot2</code></em> package available options.</li>
</ol>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="09_core_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>


</section>
</section>


<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>Cantoran et al.&nbsp;2023 DOI:<a href="https://doi.org/10.1093/femsec/fiad098" class="uri">https://doi.org/10.1093/femsec/fiad098</a><a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>2024, Jelmer Poelstra</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/jelmerp/pracs-sp24">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>